### PR TITLE
conv3d_transpose#1793

### PR DESCRIPTION
**Pull request by DankGunther**
I updated the conv3d_transpose function for Jax -  ivy.functional.backends.jax.layers

conv3d_transpose already has a backend implementation for mxnet, torch and Tensorflow so I didn't change anything in them for this task.

I didn't make any changes for this function in numpy because I could not find a way to implement it correctly.

Note: Please let me know if I implemented this function in the Jax backend properly. As well as any other changes that would need to be made

### CODE REVIEW

**Comment by DankGunther** [comment]
@Auxia please could you review this 
Thank you

**Comment by DankGunther** [comment]
@djl11 Please can I have someone else assigned to my pull request?
I haven't got a response on this for 2 days now

**Comment by DankGunther** [comment]
Hi @Auxia Please could you review this or maybe someone else can be assigned to this pull request

File: ivy/functional/backends/jax/layers.py [hunk 1]
```diff
@@ -60,5 +60,15 @@ def conv3d(*_):
...
-    raise Exception("Convolutions not yet implemented for jax library")
+def conv3d_transpose(
+    x: JaxArray,
+    filters: JaxArray,
+    strides: Union[int, Tuple[int], Tuple[int, int], Tuple[int, int, int]],
+    padding: str,
+    data_format: str = "NDHWC",
+    dilations: int = 1,
+) -> JaxArray:
+    return JaxArray.nn.conv3d_transpose(
```
Comment (juliagsy) [inline ivy/functional/backends/jax/layers.py:71]: this seems to be invalid in Jax, maybe you can try implementing this function through jax.lax.convd_transpose? Reference here: https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.conv_transpose.html 

Some additional code may be needed as well, you might want to check out the its implementation in mxnet or torch for some insights!

**Comment by juliagsy** [comment]
Hi! Thanks for contributing! I'll make a review soon!

File: ivy/functional/backends/jax/layers.py [hunk 2]
```diff
@@ -60,5 +60,15 @@ def conv3d(*_):
...
+        x, filters, strides, padding, data_format, dilations
+    )
+
```
Comment (juliagsy) [inline ivy/functional/backends/jax/layers.py:74]: For NumPy, I'll try to help searching up some information and get back to you soon!

**Review by juliagsy — CHANGES_REQUESTED** [review]
Thanks for contributing! Please refer to the comments for changes and request my review once you're done!

File: ivy/functional/backends/jax/layers.py (hunk 1 above)
Comment (DankGunther) [reply ivy/functional/backends/jax/layers.py:71]: Just a quick question if you don't mind. 
I just need to make corrections to line 68-74, right?

And also, do I need to use all the parameters mentioned in the documentation?

File: ivy/functional/backends/jax/layers.py (hunk 2 above)
Comment (DankGunther) [reply ivy/functional/backends/jax/layers.py:74]: Thank you so much for your response.
I really appreciate it

File: ivy/functional/backends/jax/layers.py (hunk 1 above)
Comment (juliagsy) [reply ivy/functional/backends/jax/layers.py:71]: yep, just the function body

and the parameters used should be the ones common to the arguments here, eg. strides, padding, dilation, etc.

Comment (DankGunther) [reply ivy/functional/backends/jax/layers.py:71]: Alright I just made changes to Jax. 
how do I commit it to this pull request?😅

**Review by DankGunther — COMMENTED** [review]
I have updated the function based on the documentation you sent me :)

File: ivy/functional/backends/jax/layers.py [hunk 3]
```diff
@@ -60,5 +60,14 @@ def conv3d(*_):
...
 
 
-def conv3d_transpose(*_):
-    raise Exception("Convolutions not yet implemented for jax library")
+def conv3d_transpose(
+    x: JaxArray,
+    filters: JaxArray,
+    strides: Union[int, Tuple[int], Tuple[int, int], Tuple[int, int, int]],
+    padding: Union[str, Sequence[Tuple[int, int]]],
+    dilations: Optional[Sequence[int]],
```
Comment (juliagsy) [inline ivy/functional/backends/jax/layers.py:68]: in our case, type hints of `dilations` should be `Union[int, Tuple[int], Tuple[int, int], Tuple[int, int, int]] = 1,`

File: ivy/functional/backends/jax/layers.py [hunk 4]
```diff
@@ -60,5 +60,14 @@ def conv3d(*_):
...
+    data_format: str = "NDHWC"
+) -> JaxArray:
+    return JaxArray.nn.conv3d_transpose(
```
Comment (juliagsy) [inline ivy/functional/backends/jax/layers.py:71]: I don't think this function is available in Jax, as it throws error when I tried running, it should make use of `jax.lax.conv_transpose`

File: ivy/functional/backends/jax/layers.py [hunk 5]
```diff
@@ -60,5 +60,14 @@ def conv3d(*_):
...
+        x, filters, strides, padding, None, dilations, (data_format, "WIO", data_format)
+    )
```
Comment (juliagsy) [inline ivy/functional/backends/jax/layers.py:73]: as `data_format` is not a default parameter in the Jax version of the function, additional code should be included to handle it, example or reference can be found in the equivalent `mxnet` and `torch` versions, links to the code are attached below!

mxnet: https://github.com/unifyai/ivy/blob/a89fc34dafa6e92b68778278fd4e6cc26e870d8e/ivy/functional/backends/mxnet/layers.py#L232

torch: https://github.com/unifyai/ivy/blob/a89fc34dafa6e92b68778278fd4e6cc26e870d8e/ivy/functional/backends/torch/layers.py#L221

**Review by juliagsy — CHANGES_REQUESTED** [review]
Hey! Looks good so far! Once the function implementation code are fixed up, you're almost good to go! Please refer to the comments above and request my review when you're done! Thank you!

File: ivy/functional/backends/jax/layers.py (hunk 3 above)
Comment (DankGunther) [reply ivy/functional/backends/jax/layers.py]: done

File: ivy/functional/backends/jax/layers.py (hunk 4 above)
Comment (DankGunther) [reply ivy/functional/backends/jax/layers.py]: You're right 
Although when I looked at the examples and documentation I found out it should be jlax.conv_transpose
I've made the changes

File: ivy/functional/backends/jax/layers.py [hunk 6]
```diff
@@ -60,5 +62,28 @@ def conv3d(*_):
...
+    strides: Union[int, Tuple[int], Tuple[int, int], Tuple[int, int, int]],
+    padding: Union[str, Sequence[Tuple[int, int]]],
+    dilations: Union[int, Tuple[int], Tuple[int, int], Tuple[int, int, int]] = 1,
+    data_format: str = "NDHWC"
+) -> JaxArray:
+    filter_shape = list(filters.shape[0:1])
+    filters = filters.permute(3, 4, 0, 1, 2)
+    if data_format == "NDHWC":
+        x = jlax.conv_transpose(0, 4, 1, 2, 3)
+    if padding == "VALID":
```
[the remark this answers is not in the document]
Comment (juliagsy) [reply ivy/functional/backends/jax/layers.py:77]: so `padding` checks are not needed

File: ivy/functional/backends/jax/layers.py [hunk 7]
```diff
@@ -1,11 +1,13 @@
 """Collection of Jax network layers, wrapped to fit Ivy syntax and signature."""
 
 # global
+import math
```
[the remark this answers is not in the document]
Comment (juliagsy) [reply ivy/functional/backends/jax/layers.py:4]: ps: this comment may be listed as the first, please refer to below comments as this is an add-on to them!

File: ivy/functional/backends/jax/layers.py [hunk 8]
```diff
@@ -60,5 +62,28 @@ def conv3d(*_):
...
+        padding_list: List[int] = [0, 0, 0]
+    elif padding == "SAME":
+        padding_list: List[int] = [math.floor(item / 2) for item in filter_shape]
+    else:
+        raise Exception(
+            "Invalid padding arg {}\n"
+            'Must be one of: "VALID" or "SAME"'.format(padding)
+        )
+    res = jlax.conv_transpose(
+        x, filters, strides, padding_list, None, dilation= dilations
```
[the remark this answers is not in the document]
Comment (juliagsy) [reply ivy/functional/backends/jax/layers.py:87]: remember to add `dim_nums` too once this instance is declared! ☺️

**Comment by juliagsy** [comment]
Hey! Sorry but I just heard from another member that they've completed the Jax implementation for this function, so would you like to continue working on this (which is fixing the docstrings in ivy/layers.py) or maybe switch to working on reformatting another function? Thanks!

File: ivy/functional/backends/jax/layers.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (DankGunther) [reply ivy/functional/backends/jax/layers.py:77]: alright thank you

File: ivy/functional/backends/jax/layers.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (DankGunther) [reply ivy/functional/backends/jax/layers.py:87]: I'll make these changes right now

File: ivy/functional/backends/jax/layers.py [hunk 9]
```diff
@@ -60,5 +62,28 @@ def conv3d(*_):
...
-    raise Exception("Convolutions not yet implemented for jax library")
+def conv3d_transpose(
+    x: JaxArray,
+    filters: JaxArray,
+    strides: Union[int, Tuple[int], Tuple[int, int], Tuple[int, int, int]],
+    padding: Union[str, Sequence[Tuple[int, int]]],
+    dilations: Union[int, Tuple[int], Tuple[int, int], Tuple[int, int, int]] = 1,
+    data_format: str = "NDHWC"
+) -> JaxArray:
+    filter_shape = list(filters.shape[0:1])
```
[the remark this answers is not in the document]
Comment (DankGunther) [reply ivy/functional/backends/jax/layers.py]: I've implemented the requested changes, please review. 
Also, maybe we could look at the numpy implementation for this function :)

[the remark this answers is not in the document]
Comment (juliagsy) [reply ivy/functional/backends/jax/layers.py]: Thanks for the changes! It's good for merging! I'll fix some lint checks and you're good to go!
